File: scanpy_hvg/critic.py
"""
scanpy_hvg critic — Evaluates HVG selection quality

FEEDBACK KEYS (must match parameter_science_guide conditions in skill.yaml):
    too_few_hvg          → n_top_genes or min_disp too restrictive
    too_many_hvg         → n_top_genes too high; noise included
    low_variance_dominated → mean dispersion of selected HVGs is low
    missing_rare_markers → pct_hvg very low relative to dataset size
"""


def critic_post_process(adata, context=None):
    context = context or {}
    protocol = context.get('protocol', '10x Genomics')
    n_cells = context.get('n_cells', adata.n_obs)

    thresholds = {
        '10x Genomics': {'min_hvg': 500,  'max_hvg': 5000, 'min_disp_mean': 0.4},
        'Smart-seq2':   {'min_hvg': 1000, 'max_hvg': 8000, 'min_disp_mean': 0.3},
        'Drop-seq':     {'min_hvg': 500,  'max_hvg': 4000, 'min_disp_mean': 0.4},
        'CITE-seq':     {'min_hvg': 1000, 'max_hvg': 6000, 'min_disp_mean': 0.35},
        'Multiome':     {'min_hvg': 1000, 'max_hvg': 6000, 'min_disp_mean': 0.3},
    }
    t = thresholds.get(protocol, thresholds['10x Genomics'])

    metrics = {}
    warnings = []
    feedback = []

    if 'highly_variable' not in adata.var.columns:
        return {
            'metrics': {},
            'warnings': ['CRITICAL: highly_variable not in adata.var — HVG step may not have run'],
            'feedback': [],
            'success': False,
            'context_used': protocol
        }

    hvg_mask = adata.var['highly_variable']
    metrics['n_hvg'] = int(hvg_mask.sum())
    metrics['n_genes_total'] = int(adata.n_vars)
    metrics['pct_hvg'] = float(metrics['n_hvg'] / metrics['n_genes_total'] * 100)

    if 'dispersions_norm' in adata.var.columns:
        metrics['mean_dispersion_hvg'] = float(
            adata.var.loc[hvg_mask, 'dispersions_norm'].mean()
        )
    else:
        metrics['mean_dispersion_hvg'] = float('nan')

    # Check HVG count bounds
    if metrics['n_hvg'] < t['min_hvg']:
        warnings.append(
            f"TOO FEW HVGs: {metrics['n_hvg']} selected (min {t['min_hvg']} for {protocol}). "
            "Consider increasing n_top_genes or decreasing min_disp."
        )
        feedback.append('too_few_hvg')

    if metrics['n_hvg'] > t['max_hvg']:
        warnings.append(
            f"TOO MANY HVGs: {metrics['n_hvg']} selected (max {t['max_hvg']} for {protocol}). "
            "Consider decreasing n_top_genes."
        )
        feedback.append('too_many_hvg')

    # Check dispersion quality
    if (not float('nan') == metrics['mean_dispersion_hvg'] and
            metrics['mean_dispersion_hvg'] < t['min_disp_mean']):
        warnings.append(
            f"LOW DISPERSION: mean normalized dispersion of HVGs = "
            f"{metrics['mean_dispersion_hvg']:.3f} (threshold {t['min_disp_mean']}). "
            "Selected genes may be weakly variable."
        )
        feedback.append('low_variance_dominated')

    return {
        'metrics': metrics,
        'warnings': warnings,
        'feedback': feedback,
        'success': not bool(feedback),
        'context_used': protocol
    }

File: seurat_labeltransfer/critic.R
#' seurat_labeltransfer critic — Evaluates label transfer quality
#'
#' FEEDBACK KEYS (match parameter_science_guide in skill.yaml):
#'   low_prediction_score  → mean score too low; try more dims or different reference
#'   noisy_transfer        → high score variance; too many dims
#'   fragmented_predictions → predictions inconsistent within clusters
#'   over_smoothed         → too few types detected; k_weight too high

critic_post_process <- function(obj, context = NULL) {
    if (is.null(context)) context <- list()
    protocol <- if (!is.null(context$protocol)) context$protocol else "10x Genomics"

    thresholds <- list(
        "10x Genomics" = list(min_pred_score = 0.5, min_high_conf = 40.0, min_types = 2L),
        "Smart-seq2"   = list(min_pred_score = 0.6, min_high_conf = 50.0, min_types = 2L),
        "Drop-seq"     = list(min_pred_score = 0.45, min_high_conf = 35.0, min_types = 2L),
        "CITE-seq"     = list(min_pred_score = 0.55, min_high_conf = 45.0, min_types = 2L)
    )
    t <- if (!is.null(thresholds[[protocol]])) thresholds[[protocol]] else thresholds[["10x Genomics"]]

    metrics  <- list()
    warnings <- character(0)
    feedback <- character(0)

    # Find prediction column
    pred_cols <- grep("predicted_celltype|predicted\\.id", colnames(obj@meta.data), value = TRUE)
    score_cols <- grep("\\.score$", colnames(obj@meta.data), value = TRUE)

    if (length(pred_cols) == 0L) {
        return(list(
            metrics      = list(),
            warnings     = "CRITICAL: No label transfer prediction columns found — run seurat_labeltransfer first",
            feedback     = character(0),
            success      = FALSE,
            context_used = protocol
        ))
    }

    labels <- obj@meta.data[[pred_cols[1]]]
    metrics$n_cells           <- as.integer(ncol(obj))
    metrics$n_predicted_types <- as.integer(length(unique(labels)))
    metrics$top3_predicted    <- names(sort(table(labels), decreasing = TRUE))[1:min(3L, metrics$n_predicted_types)]

    if (length(score_cols) > 0) {
        scores <- obj@meta.data[[score_cols[1]]]
        metrics$mean_prediction_score <- mean(scores, na.rm = TRUE)
        metrics$pct_high_confidence   <- mean(scores >= 0.75, na.rm = TRUE) * 100
        metrics$sd_prediction_score   <- sd(scores, na.rm = TRUE)

        if (metrics$mean_prediction_score < t$min_pred_score) {
            warnings <- c(warnings, sprintf(
                "LOW PREDICTION SCORE: mean = %.3f (threshold %.2f for %s). "
                "Reference may not match tissue. Try increasing dims.",
                metrics$mean_prediction_score, t$min_pred_score, protocol
            ))
            feedback <- c(feedback, "low_prediction_score")
        }

        if (metrics$pct_high_confidence < t$min_high_conf) {
            warnings <- c(warnings, sprintf(
                "FEW HIGH-CONFIDENCE PREDICTIONS: %.1f%% cells with score ≥ 0.75 "
                "(threshold %.0f%%). Reference may be mismatched.",
                metrics$pct_high_confidence, t$min_high_conf
            ))
            if (!"low_prediction_score" %in% feedback) {
                feedback <- c(feedback, "low_prediction_score")
            }
        }

        # Check prediction consistency within clusters
        cluster_cols <- intersect(c("seurat_clusters", "leiden", "louvain"), colnames(obj@meta.data))
        if (length(cluster_cols) > 0) {
            cluster_labels <- table(obj@meta.data[[cluster_cols[1]]], labels)
            cluster_entropy <- apply(cluster_labels, 1, function(row) {
                p <- row / sum(row)
                p <- p[p > 0]
                -sum(p * log(p))
            })
            if (mean(cluster_entropy) > 1.5) {
                warnings <- c(warnings,
                    "FRAGMENTED PREDICTIONS: High label entropy within clusters. "
                    "Consider increasing k_weight for smoother predictions."
                )
                feedback <- c(feedback, "fragmented_predictions")
            }
        }
    }

    if (metrics$n_predicted_types < t$min_types) {
        warnings <- c(warnings, sprintf(
            "TOO FEW CELL TYPES: %d type(s) predicted. k_weight may be too high.",
            metrics$n_predicted_types
        ))
        feedback <- c(feedback, "over_smoothed")
    }

    list(
        metrics      = metrics,
        warnings     = warnings,
        feedback     = feedback,
        success      = length(feedback) == 0L,
        context_used = protocol
    )
}
